src/trajectory.jl

```diff
@@ -565,19 +575,37 @@ function isterminated(h::Hamiltonian, t::BinaryTree{<:ClassicNoUTurn})
 end
 
 """
-    isterminated(h::Hamiltonian, t::BinaryTree{<:GeneralisedNoUTurn})
+    isterminated(
+        h::Hamiltonian, 
+        t::BinaryTree{<:GeneralisedNoUTurn},
+        tleft::BinaryTree{<:GeneralisedNoUTurn}, 
+        tright::BinaryTree{<:GeneralisedNoUTurn}
+    )
 
 Detect U turn for two phase points (`zleft` and `zright`) under given Hamiltonian `h`
 using the generalised no-U-turn criterion.
 
-Ref: https://arxiv.org/abs/1701.02434
+Ref: https://arxiv.org/abs/1701.02434 https://github.com/stan-dev/stan/pull/2800
 """
-function isterminated(h::Hamiltonian, t::BinaryTree{<:GeneralisedNoUTurn})
-    # z0 is starting point and z1 is ending point
-    z0, z1 = t.zleft, t.zright
+function isterminated(
+    h::Hamiltonian, 
+    t::BinaryTree{<:GeneralisedNoUTurn}, 
+    tleft::BinaryTree{<:GeneralisedNoUTurn}, 
+    tright::BinaryTree{<:GeneralisedNoUTurn}
+)
     rho = t.c.rho
-    s = (dot(rho, ∂H∂r(h, -z0.r)) >= 0) || (dot(-rho, ∂H∂r(h, z1.r)) >= 0)
-    return Termination(s, false)
+    s1 = generalised_uturn_criterion(rho, ∂H∂r(h, t.zleft.r), ∂H∂r(h, t.zright.r))
```

Comment:
Do we still need this check?

Reply:
Yes, as I understand the new checks don't catch all U-turns by themselves. Also both Stan (https://github.com/stan-dev/stan/blob/0b92731851947f91d055a61d922b491eb3b14f01/src/stan/mcmc/hmc/nuts/base_nuts.hpp#L178) and PyMC3 (https://github.com/pymc-devs/pymc3/blob/90f48ed22e3c28652903fa27aaaeff59aa29acdf/pymc3/step_methods/hmc/nuts.py#L314) still have this check.

Reply:
I see. Thanks for clarifying this.